### PR TITLE
ARROW-16327: [Java][CI] Add Java 17 to CI matrix for java workflows

### PR DESCRIPTION
**Pull request by raulcd**
This PR aims to add support for java 17 on CI as required on https://github.com/apache/arrow/pull/12941
We probably should cherrypick the commit here on that PR.

### CODE REVIEW

**Comment by raulcd** [comment]
@github-actions crossbow submit *java*

**Review by lidavidm — APPROVED** [review]
Thanks for setting this up!

**Comment by kou** [comment]
It seems that `maven:3.8.5-openjdk-17` is based on Oracle Linux 8 not Debian:

```console
$ docker run --rm maven:3.8.5-openjdk-17 cat /etc/system-release
Oracle Linux Server release 8.5
```

So existing `ci/docker/debian-9-java.dockerfile` filename isn't suitable for OpenJDK 17.

How about adding `ci/docker/oracle-8-java.dockerfile`, adding `oracle-java` to `docker-compose.yml` and removing `jdk_name` from `.env`?

BTW, it seems that nobody uses `ARROW_JAVA_SHADE_FLATBUFS` defined in `ci/docker/debian-9-java.dockerfile`. See also: 7bc2b0f3579b6fc2d1d6d9797f3a1e5e3726131f

We may be able to use `maven:*` images directly instead of defining our (thin) wrapper image. For example, we use existing images directly for `python-wheel-manylinux-test-imports`.

**Comment by kou** [comment]
Could you also confirm CI failure? https://github.com/apache/arrow/runs/6218109762?check_suite_focus=true

```text
Error:  Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.6.2:compile (default-compile) on project arrow-format: Compilation failure -> [Help 1]
Error:  Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.6.2:compile (default-compile) on project arrow-plasma: Compilation failure -> [Help 1]
Error:  Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.6.2:compile (default-compile) on project arrow-memory-core: Compilation failure -> [Help 1]
```

The following log may be related:

```text
[INFO] -----------------< org.apache.arrow:arrow-memory-core >-----------------
[INFO] Building Arrow Memory - Core 8.0.0-SNAPSHOT                       [6/19]
[INFO] --------------------------------[ jar ]---------------------------------
Warning:  *****************************************************************
Warning:  * Your build is requesting parallel execution, but project      *
Warning:  * contains the following plugin(s) that have goals not marked   *
Warning:  * as @threadSafe to support parallel building.                  *
Warning:  * While this /may/ work fine, please look for plugin updates    *
Warning:  * and/or request plugins be made thread-safe.                   *
Warning:  * If reporting an issue, report it against the plugin in        *
Warning:  * question, not against maven-core                              *
Warning:  *****************************************************************
Warning:  The following plugins are not marked @threadSafe in Arrow Memory - Core:
Warning:  org.apache.rat:apache-rat-plugin:0.13
Warning:  Enable debug to see more precisely which goals are not marked @threadSafe.
Warning:  *****************************************************************
```

**Comment by raulcd** [comment]
Thanks for the review @kou !  I have added a new `oracle-java` entry on the `docker-compose.yml` and I have updated the `debian-java` to use the `maven:*` images directly removing the wrapper image `ci/docker/debian-9-java.dockerfile`, as you said I can't find anywhere where the `ARROW_JAVA_SHADE_FLATBUFS` were being used. I have also removed the `JDK_NAME` and have added a JDK 8 CI test (as it seems to be the default defined on our .env).

As per the failure on JDK 17 it is expected until we merge https://github.com/apache/arrow/pull/12941, the idea is to rebase once that PR is merged and validate the JDK 17 image is passing at that point!

File: docker-compose.yml
Comment (pitrou) [inline docker-compose.yml:1485]: How is this different from `debian-java`? It seems to be based on the same base image.

Comment (raulcd) [reply docker-compose.yml:1485]: the maven image for openjdk 17 is based on Oracle Linux instead of debian `maven:3.8.5-openjdk-17`:
```
$ docker run --rm maven:3.8.5-openjdk-17 cat /etc/system-release
Oracle Linux Server release 8.5
```

**Comment by kou** [comment]
@github-actions crossbow submit java-jars

**Comment by raulcd** [comment]
@kou @pitrou is there anything else you think we should do on this PR in order to get it merged?
About the `oracle-java` vs `debian-java` I am not sure how we could manage the inconsistency from the maven images repository better on our side.I did add the params to the usage on the `docker-compose` and the `openjdk` vs `jdk` naming convention for the image:
```
    # Usage:
    #   docker-compose build oracle-java
    #   docker-compose run oracle-java
    # Parameters:
    #   MAVEN: 3.8.5
    #   JDK: 17
    image: ${ARCH}/maven:${MAVEN}-openjdk-${JDK}
```
vs
```
    # Usage:
    #   docker-compose build debian-java
    #   docker-compose run debian-java
    # Parameters:
    #   MAVEN: 3.5.4, 3.6.2
    #   JDK: 8, 11
    image: ${ARCH}/maven:${MAVEN}-jdk-${JDK}
```
let me know if you have a better idea.

**Comment by pitrou** [comment]
@raulcd I think this is good to go, I'll let @kou comment if he wants to.

**Review by kou — APPROVED** [review]
+1

Thanks!

**Pull request closed without merging**
